javascripts/2048.js

```diff
@@ -1,32 +1,112 @@
-  console.log('ready!');
-  $('body').keydown(function(event){
-    var arrow_keys = [37, 38, 39, 40];
-    if(arrow_keys.indexOf(event.which) > -1) {
-      var tile = $('.tile');
-      moveTile(tile, event.which);
-      event.preventDefault();
+  initializeGame();
+  $('body').keydown(playTurn);
+  $('.retry-button').click(retryClickHandler);
+});
+
+function initializeGame() {
+  // randomly pick two positions and start values
+  addTile(newTileValue);
+  addTile(newTileValue);
+}
+
+function playTurn(event) {
+  if (gameOver) { return; }
+
+  var arrow_keys = [37, 38, 39, 40];
+  if (arrow_keys.indexOf(event.which) > -1) {
+    event.preventDefault();
+
+    var collection = new TileCollection(event.which);
+    collection.move();
+
+    if ($('.tile').length >= boardSize) {
+      if (checkLose()) { endGame('lose'); }
+    } else {
+      addTile(newTileValue);
     }
-  })
-})
-
-function moveTile(tile, direction) {
-  var new_tile_value = tile.attr("data-val") * 2;
-  tile.attr("data-val", new_tile_value);
-  tile.text(new_tile_value);
-
-  switch(direction) {
-    case 38: //up
-      tile.attr("data-row","r0");
-      break;
-    case 40: //down
-      tile.attr("data-row","r3");
-      break;
-    case 37: //left
-      tile.attr("data-col","c0");
-      break;
-    case 39: //right
-      tile.attr("data-col","c3");
-      break;
+
+    if (!priorWin) { checkWin(); }
+  }
+}
+
+function retryClickHandler(event) {
+  event.preventDefault();
+  location.reload();
+}
+
+function continueClickHandler(event) {
+  event.preventDefault();
+  gameOver = false;
+  $('.gameOver, .cont-button, .retry-button').remove();
+}
+
+function addTile(possible_values) {
+  var tile = $('<div class="tile" data-row="" data-col="" data-val="">');
+  var randomRow = function() { return Math.floor(Math.random() * (rows.length)); }();
+  var randomCol = function() { return Math.floor(Math.random() * (cols.length)); }();
+  var randomValue = function() {
+    return Math.floor(Math.random() * (possible_values.length));
+    }();
+
+  tile.attr('data-row', rows[randomRow]);
+  tile.attr('data-col', cols[randomCol]);
+  tile.attr('data-val', possible_values[randomValue]);
+  tile.text(tile.attr('data-val'));
+
+  if ($('[data-row=' + rows[randomRow] + '][data-col=' + cols[randomCol] + ']').length === 0) {
+
+    $('#gameboard').append(tile);
+    scaleIn(tile);
+  } else {
+    addTile(possible_values);
+  }
+}
+
+function updateValue(tile) {
+  var added = tile.attr('data-val') * 2;
+  tile.attr('data-val', added);
+  tile.text(added);
+  pop(tile);
+
+  updateScore(added);
+}
+
+function updateScore(num) {
+  score += num;
+  $('.score').text(score);
+  animateAddedValue(num);
+}
+
+// Sequentially collects all existing tiles from a row.
+function generateRow(num) {
+  var row = [];
+  for (var i = 0; i < rows.length; i++) {
+    var tile = $('[data-row=' + num + '][data-col=' + i + ']');
+    if (tile.length > 0) { row.push(tile); }
```

Comment:
A little trick you can use with javascript arrays is `concat`.
```
var tile = $('[data-row=' + num + '][data-col=' + i + ']');
row = row.concat(tile);
```
This should work, as jquery selectors always return an array, even if the selector finds no qualifying items.